WickedEngine/wiHelper.cpp

```diff
@@ -112,6 +112,197 @@ namespace wi::helper
 		}
 	}
 
+	bool loadTextureFromMemory(wi::vector<uint8_t>& texturedata, const wi::graphics::Texture& texture)
+	{
+		using namespace wi::graphics;
+
+		GraphicsDevice* device = wi::graphics::GetDevice();
+
+		TextureDesc desc = texture.GetDesc();
+
+		uint32_t data_count = desc.width * desc.height;
+		uint32_t data_stride = wi::graphics::GetFormatStride(desc.format);
+		uint32_t data_size = data_count * data_stride;
+
+		Texture stagingTex;
+		TextureDesc staging_desc = desc;
+		staging_desc.usage = Usage::UPLOAD;
+		staging_desc.layout = ResourceState::COPY_SRC;
+		staging_desc.bind_flags = BindFlag::NONE;
+		staging_desc.misc_flags = ResourceMiscFlag::NONE;
+		staging_desc.mip_levels = 1;
+		bool success = device->CreateTexture(&staging_desc, nullptr, &stagingTex);
+		assert(success);
+
+		if (stagingTex.mapped_data != nullptr )
```

Comment:
This texture copy is not very robust, it will fail for textures except simple 2D textures. The `saveTextureToMemory` function has an example of a correct texture data copy that will correctly handle mip levels, slices and 3D textures. The copy must take into account the row and slice pitch to work correctly. It also handles block compressed texture format correctly.

For simpler implementation, you can also use the `CreateTexture()` second argument initial data, and you don't need to invoke copies on command list. (if that fits your needs)